zingo_wallet/src/interface.rs

```diff
@@ -0,0 +1,191 @@
+}
+
+impl zcash_wallet_interface::Wallet for ZingoWallet {
+    fn user_agent_id() -> zcash_wallet_interface::UserAgentId {
+        const AGENT_NAME: &str = "zingo_wallet";
+        const VERSION: &str = "0.0.1";
+
+        zcash_wallet_interface::UserAgentId {
+            name: AGENT_NAME.to_string(),
+            version: VERSION.to_string(),
+        }
+    }
+
+    async fn new_wallet() -> Self {
+        // we cannot instantiate the current version of the lightclient yet
+        // without assumptions about keys and servers
+        // which would violate principles of the interface
+        // so we dont
+        ZingoWallet {
+            keys: Vec::new(),
+            lightclient: None,
+        }
+    }
+
+    type AddServerError = AddServerError;
+
+    async fn add_server(&mut self, server_address: String) -> Result<(), Self::AddServerError> {
+        use std::num::NonZeroU32;
+
+        use std::str::FromStr as _;
+        use std::sync::Arc;
+        use std::sync::RwLock;
+
+        use zingolib::config::ChainType;
+        use zingolib::config::SyncConfig;
+        use zingolib::config::TransparentAddressDiscovery;
+
+        use zingolib::config::ZingoConfigBuilder;
+        use zingolib::config::chain_from_str;
+        use zingolib::lightclient::LightClient;
+        use zingolib::wallet::LightWallet;
+        use zingolib::wallet::WalletSettings;
+        if self.keys.len() == 1
+            && let Some(key) = self.keys.first()
+        {
+            let server_uri = Uri::from_str(server_address.as_str())
+                .map_err(|invalid_uri| AddServerError::CantParseUri(server_address, invalid_uri))?;
+
+            let lightwalletd_uri: Arc<RwLock<Uri>> = Arc::new(RwLock::new(server_uri.clone()));
+
+            let (chain_type, birthday) = {
+                // we need to ask the indexer for this information
+
+                let mut client = {
+                    // global configuration must be manually set *somewhere*
+                    rustls::crypto::ring::default_provider().install_default();
+                    zingolib::grpc_client::get_zcb_client(server_uri.clone())
+                        .await
+                        .map_err(|e| AddServerError::CantCreateClient(server_uri.clone(), e))?
+                };
+
+                let lightd_info = client
+                    .get_lightd_info(tonic::Request::new(
+                        zcash_client_backend::proto::service::Empty {},
+                    ))
+                    .await?
+                    .into_inner();
+
+                let chain_name = &lightd_info.chain_name;
+                let chain_type: ChainType = chain_from_str(chain_name)?;
+
+                let birthday: zcash_primitives::consensus::BlockHeight =
+                    lightd_info.block_height.try_into()?;
+                (chain_type, birthday)
+            };
+
+            // this seems like a lot of set up. Do we really need all this right here??
```

Comment:
It seems to me that the wallet-related functionality defined from here to the end of this fn might be useful outside the context of server addition.

So why not factor this out as another fn?